### ISSUE TITLE
unblacklist gulp-rollup

### ISSUE THREAD

**Issue by flying-sheep** [gulpjs/plugins#205]
hi, after struggling to use rollup directly and seeing that rollup-stream is incompatible with gulp-plumber i wonder why this is blacklisted.

i *much* rather do 

```js
return gulp.src('src/js/main.js')
	.pipe(plumber())
	.pipe(rollup({ sourceMap: true }))
	.pipe(sourcemaps.init({ loadMaps: true }))
	.pipe(uglify())
	.pipe(sourcemaps.write('.'))
	.pipe(gulp.dest('js/'))
```

than

```js
const stream = rollupStream({
	entry: 'src/js/main.js',
	sourceMap: true,
})
return stream
	.on('error', e => {
		console.error(`${e.stack}`)
		stream.emit('end')
	})
	.pipe(source('src/js/main.js'))
	.pipe(buffer())
	.pipe(sourcemaps.init({ loadMaps: true }))
	.pipe(uglify())
	.pipe(sourcemaps.write('.'))
	.pipe(gulp.dest('js/'))
```

or whatever arcane incantations you have to go through to mace the recommended way work.

**Comment by phated** [comment]
It is against gulp guidelines to use gulp.src as a globber for a bundler (as it reads from the filesystem).  The blacklist are a strong suggestion by the gulp team not to use certain plugins and indicates we won't help with any problems that may be caused by them.  That being said, you are free to use what you want.

**Comment by flying-sheep** [comment]
> as a globber for a bundler

i’m not sure i understand – how would a in-stream bundler have to work to be removed from the blacklist?

* simply used the one source file streamed in via gulp.src?
* should it provide a “globber” that resolves dependencies and returns a vinyl stream of input files which can be transformed before really being bundled?
* another way?

----

that being said, the blacklist reason is a bit strange – using rollup directly will give me a promise and i don’t know how to elegantly start a gulp pipeline from that. rollup-stream is maybe a better suggestion there

**Comment by Permutatrix** [comment]
@flying-sheep

>should it provide a “globber” that resolves dependencies and returns a vinyl stream of input files which can be transformed before really being bundled?

Something similar [was discussed](https://github.com/mcasimir/gulp-rollup/pull/16#issuecomment-156668598) a while back. I later made rollup-stream and considered the problem solved, but I did think the concept was interesting. Providing our own globber is more interesting still, and it makes the syntax very neat, too. I think it's an excellent idea. Of course, the option of using a different globber and specifying the entry file at the point where it's bundled should be left open.

You've shown that this way of using gulp-rollup has more merit than I thought. It's 1 a.m. where I am right now, but I'd definitely be happy to get started on implementing this tomorrow as v2.0.0. Unless I'm once again being blinded to the zen of gulp&mdash;@phated, if you wouldn't mind giving an assessment of whether this would get gulp-rollup unblacklisted, I'd appreciate it.

**Comment by flying-sheep** [comment]
> using a different globber and specifying the entry file at the point where it's bundled should be left open.

my idea was that the stream gets amended with metadata about dependencies, because that avoids redundant work (infer dependencies via file system → emit metadata-enhanced file list → infer dependencies within file list)

but obviously `gulp.src('src/js/**.js').pipe(rollup('src/js/main.js'))...` should also work and simply result in the dependencies being inferred in a later step.

**Comment by Permutatrix** [comment]
@flying-sheep That may not be a good idea, let alone possible. What would happen if a file was modified between `rollup.src` and `rollup.bundle` and its dependencies changed?

Another thing to consider is that Rollup parses files, and therefore emits errors, during the dependency discovery stage. This scheme is good for other reasons, but gulp-plumber still wouldn't be able to do anything about it if Rollup had an error, just like with rollup-stream. Using `gulp.src` would be a different story, though.

**Comment by Permutatrix** [comment]
@flying-sheep Although... it might work if you put gulp-plumber first, like so:

```js
plumber()
    .pipe(rollup.src('src/js/main.js'))
    .pipe(sourcemaps.init())
    .pipe(rollup.bundle())
    .pipe(uglify())
    .pipe(sourcemaps.write('.'))
    .pipe(gulp.dest('js/'))
```

Just a crazy thought. It's too late at night for me to be having thoughts of any other kind.

**Comment by flying-sheep** [comment]
you’re right in both things.

still, even with that limitation, not using rollup as mandatory entry point is more gulpy than rollup-stream.

maybe we can do an optimization which attaches the metadata and uses it when the files weren’t modified since…

and does the “putting plumber first” actually work? well, we won’t need plumber anymore with gulp 4 so it’s just a matter of time.

**Comment by flying-sheep** [comment]
so this was really fixed in #218, yay!
